### PR TITLE
Documentation Overhaul

### LINKED ISSUE #1565: Inheritance of `pyvista.PolyData`, error by `.overwrite()`
In our project we heavily rely on the `pyvista.PolyData` object, therefore we made our own class - which inherits from `pyvista.PolyData` - with our own custom methods.

However when using functions with the `inplace` argument, the underlying function `overwrite` causes an AssertionError because the types doesn't match.

Code to reproduce at Windows, PyVista version 0.31.1
```python
import pyvista
import numpy

class Foo(pyvista.PolyData):
    pass

foo = Foo()
foo.points = np.eye(3)
foo.delaunay_2d(inplace=True)


Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\git\geams-lib\.venv39\lib\site-packages\pyvista\core\filters.py", line 4971, in delaunay_2d
    poly_data.overwrite(mesh)
  File "C:\git\geams-lib\.venv39\lib\site-packages\pyvista\core\dataset.py", line 767, in overwrite
    raise TypeError('The Input DataSet type must match '
TypeError: The Input DataSet type must match the one being overwritten <class '__main__.Foo'>
```

We have a current workaround where we monkey-patch the original `PolyData` class, but this is farm from neat or ideal.

Proposal:
Maybe loosen the type-checking, or use `issubclass` instead of `isinstance`.

### PR DESCRIPTION
**Pull request by akaszynski**
Addresses several documentation related request (including #1531).

### Data Model Documentation Checklist
- [x] Creating Arrays from VTK, numpy, and list.
- [x] Creating geometry from those arrays.
- [x] Data arrays
  - [x] Point data
  - [x] Cell data
  - [x] Field data
- [x] Data attributes and methods.
- [x] Other geometry types?
- [x] Comparison of vtk `__str__` and pyvista `__repr__`

### Misc
- [x] Add PyVista vs. VTK section
- [x] Cleanup what is a mesh
- [x] Use templates for the rest of our class APIs

Additionally:
- Several examples were added to improve our example code coverage.

### CODE REVIEW

**Comment by adeak** [comment]
Let me just note that `issubclass` is a red herring here. Looking at
```diff
-        if not isinstance(mesh, type(self)):
-            raise TypeError('The Input DataSet type must match '
+        # using subclass here to allow users to subclass our datatypes
+        if not issubclass(type(self), type(mesh)):
```
the real change is the reversal of the arguments. `isinstance` already allows the first argument to be a subclass of the second, which is exactly why `isinstance` should be used instead of `type(this) is type(that)` (all in the name of duck typing):
```py
class Foo:
    pass

class Bar(Foo):
    pass

print(isinstance(Bar(), Foo)) # True
```

So using `issubclass(type(...), ...)` instead of `isinstance` is an anti-pattern. Just replace the order of arguments, which should be the real change here:
```diff
-        if not isinstance(mesh, type(self)):
-            raise TypeError('The Input DataSet type must match '
+        if not isinstance(self, type(mesh)):
```

And I guess we should ask ourselves what this reversal of argument really implies, why it makes sense, and write _that_ reasoning into a comment if necessary.

**Comment by akaszynski** [comment]
@adeak, good point.  Fixing...

**Comment by akaszynski** [comment]
@adeak, in standard Alex style, I'm expanding the focus of this PR to resolve #1531.  It would be helpful if you could weigh in while I'm writing it.

@adamgranthendry, your inputs would be helpful too.

**Comment by adamgranthendry** [comment]
@akaszynski Happy to help! Would you like me to try to add my notes from [How to Work with Data Types #467](https://github.com/pyvista/pyvista-support/issues/467) to a restructured text file for the docs? What's the best format to use here?

**Comment by akaszynski** [comment]
For the moment, I'd like feedback regarding `data_model.rst` in this PR.  Add in the comments of this PR anything you'd like me to add or change.

**Comment by adamgranthendry** [comment]
@akaszynski Do you have a timeline you're trying to meet as well so I can get you feedback in a timely fashion?

**Comment by akaszynski** [comment]
It's going to be a WIP today as I write this down.  Right now I say I'm about 25% done.  Adding a checklist above.

File: doc/user-guide/data_model.rst
Comment (adamgranthendry) [inline doc/user-guide/data_model.rst:25]: A picture is worth a thousand words. Can we add:

image

Also, can we change this paragraph to:

"Geometry in PyVista is represented as points and cells. As a simple example, consider the following 5x5 array:

```python
values = np.array([
    [1, 0, 0, 1, 1],
    [0, 1, 0, 0, 1],
    [1, 1, 0, 0, 0],
    [1, 0, 1, 0, 1],
    [0, 1, 1, 1, 0]])
```

We clearly see this dataset consists only of 1's and 0's, but where are they positioned in space? If we assume the data lives in a `UniformGrid`, then the (x, y) positions can be given by the indeces of the array. For example, point (3, 4) (NOTE: arrays are 0-indexed) is 1 and point (2, 3) is 0. But what if these datapoints aren't all in the plane and aren't uniformly spaced apart? We would need a way to describe the position of each of these points in space, but we're limited to expressing the values themselves as we've done above (lists of arrays with indeces). `VTK` (and hence `pyvista`) have multiple classes that represent different data shapes. The most important dataset classes are shown below:

image

Here, the datasets are listed from most to least complex. That is, every dataset can be represented as an `UnstructuredGrid`, but `UnstructuredGrid`s take the most amount of memory to store since they must account for every data point individually. On the other hand, since `vtkImageData` (`UnformGrid` data) is uniformly spaced, a few integers and floats can describe the shape, so it takes the least amount of memory to store."

File: doc/user-guide/data_model.rst
Comment (adamgranthendry) [inline doc/user-guide/data_model.rst:189]: I would write "mesh connectivity / topology", since topology is used in other docs and in mathematical contexts, so it's good to know both terms.

File: doc/user-guide/data_model.rst
Comment (adamgranthendry) [inline doc/user-guide/data_model.rst:36]: Also, in here, can you demo printing out a `pyvista` and/or `vtk` dataset and quickly showing what the information means? There's quite a lot more detail here than in your typical class `__repr__()`, which can be overwhelming to the newcomer.

e.g. If I make a `vtkImageData` dataset called `id` and print it out with `printi(id)`, I get this massive list like this (I think `pyvista` has shrunk this considerably, which is good, but it would do well to have a comparison between the printout of a raw `vtk` object and wrapped `pyvista` one:

```
vtkImageData (0000023932D73E80)
  Debug: Off
  Modified Time: 126
  Reference Count: 1
  Registered Events: (none)
  Information: 0000023934B751D0
  Data Released: False
  Global Release Data: Off
  UpdateTime: 0
  Field Data:
    Debug: Off
    Modified Time: 79
    Reference Count: 1
    Registered Events: (none)
    Number Of Arrays: 0
    Number Of Components: 0
    Number Of Tuples: 0
  Number Of Points: 250000
  Number Of Cells: 249001
  Cell Data:
    Debug: Off
    Modified Time: 82
    Reference Count: 1
    Registered Events:
      Registered Observers:
        vtkObserver (0000023932E88F50)
          Event: 33
          EventName: ModifiedEvent
          Command: 0000023934B75270
          Priority: 0
          Tag: 1
    Number Of Arrays: 0
    Number Of Components: 0
    Number Of Tuples: 0
    Copy Tuple Flags: ( 1 1 1 1 1 0 1 1 1 1 1 )
    Interpolate Flags: ( 1 1 1 1 1 0 0 1 1 1 1 )
    Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Scalars: (none)
    Vectors: (none)
    Normals: (none)
    TCoords: (none)
    Tensors: (none)
    GlobalIds: (none)
    PedigreeIds: (none)
    EdgeFlag: (none)
    Tangents: (none)
    RationalWeights: (none)
    HigherOrderDegrees: (none)
  Point Data:
    Debug: Off
    Modified Time: 126
    Reference Count: 1
    Registered Events:
      Registered Observers:
        vtkObserver (0000023932E89370)
          Event: 33
          EventName: ModifiedEvent
          Command: 0000023934B75270
          Priority: 0
          Tag: 1
    Number Of Arrays: 1
    Array 0 name = my array
    Number Of Components: 1
    Number Of Tuples: 250000
    Copy Tuple Flags: ( 1 1 1 1 1 0 1 1 1 1 1 )
    Interpolate Flags: ( 1 1 1 1 1 0 0 1 1 1 1 )
    Pass Through Flags: ( 1 1 1 1 1 1 1 1 1 1 1 )
    Scalars:
      Debug: Off
      Modified Time: 76
      Reference Count: 2
      Registered Events: (none)
      Name: my array
      Data type: double
      Size: 250000
      MaxId: 249999
      NumberOfComponents: 1
      Information: 0000000000000000
      Name: my array
      Number Of Components: 1
      Number Of Tuples: 250000
      Size: 250000
      MaxId: 249999
      LookupTable: (none)
    Vectors: (none)
    Normals: (none)
    TCoords: (none)
    Tensors: (none)
    GlobalIds: (none)
    PedigreeIds: (none)
    EdgeFlag: (none)
    Tangents: (none)
    RationalWeights: (none)
    HigherOrderDegrees: (none)
  Bounds:
    Xmin,Xmax: (0, 499)
    Ymin,Ymax: (0, 499)
    Zmin,Zmax: (0, 0)
  Compute Time: 127
  Spacing: (1, 1, 1)
  Origin: (0, 0, 0)
  Direction: (1, 0, 0, 0, 1, 0, 0, 0, 1)
  Dimensions: (500, 500, 1)
  Increments: (0, 0, 0)
  Extent: (0, 499, 0, 499, 0, 0)
```

**Review by adamgranthendry — CHANGES_REQUESTED** [review]
@akaszynski I've add my first couple comments here. Please feel free to use what you like! I'll try and go through the rest of the docs tomorrow.

File: doc/user-guide/data_model.rst
Comment (adeak) [inline doc/user-guide/data_model.rst:22]: "surface or volume" sounds like something a lot more specific than what you mean. It implies connectivity. We literally just have a subset of points embedded in 3D space at this point, right?

File: doc/user-guide/data_model.rst
Comment (adeak) [inline doc/user-guide/data_model.rst:129]: If we support any sequence (which we probably do), we should probably say so. Perhaps only in parentheses, not to scare off users who don't know what that means, but this information could still benefit more savvy users (by noting that they don't have to transform their tuples into lists, for instance).

File: doc/user-guide/data_model.rst
Comment (adeak) [inline doc/user-guide/data_model.rst:167]: There's no arithmetic going on (hopefully), so this should be true with `np.array_equal`, which I'd slightly prefer here.

**Review by adeak — COMMENTED** [review]
Just a few more general remarks looking at `data_model.rst`.

@akaszynski you can ping me when you're done with a first draft and I'll give it a thorough round of copy-editing/proofreading ;)

File: doc/user-guide/data_model.rst
Comment (adeak) [reply doc/user-guide/data_model.rst:36]: @adamgranthendry just a small correction: it's usually `__str__` that is verbose and `__repr__` is short and barely informative in VTK/PyVista. It took me a pretty long time to find the verbose output, because I always rely on evaluation in the REPL, which uses `__repr__` :)

Comment (adamgranthendry) [reply doc/user-guide/data_model.rst:36]: @adeak Ah, excellent point! Yes, thanks for that!

File: doc/user-guide/data_model.rst
Comment (adamgranthendry) [reply doc/user-guide/data_model.rst:22]: @adeak Agreed. Specifically at the most fundamental level, we actually have the `vtkDataObject`, which is nothing more than `vtkFieldData` and `vtkFieldArray`s, but they are the simplest thing that can be passed to filters in `VTK`. A `vtkDataSet` expands upon `vtkDataObject`s, has field data, but also has **geometry** (notion of where the data lives in 3-space) and **topology** (notion of how the data is connnected). In mathematical terms, the different subclasses of `vtkDataSet`s are "n-simplicial complexes", typically 2-simplicial complexes for surfaces and 3-simplicial complexes for volumetric data. If the data does not need to be cleaned up and is a proper immersion in 3-space, then we say it is also "manifold".

File: doc/user-guide/data_model.rst
Comment (adamgranthendry) [reply doc/user-guide/data_model.rst:129]: @adeak Agreed. both `lists` and `tuples` (and numpy arrays made of `lists` and `tuples` are accepted). Could we be explicit and say "sequences (i.e. lists or tuples)" for the less-initiated pythonistas? :)

File: doc/user-guide/data_model.rst
Comment (akaszynski) [reply doc/user-guide/data_model.rst:23]: This is great!  I'm adding two sections to the user guide.  One is understanding how pyvista organizes data for new users who want to start working with data, and the other is for seasoned experts in VTK.

See: 01cec98b5f7c685e91bfaf611b3ecdf52833c072

File: doc/user-guide/data_model.rst
Comment (akaszynski) [reply doc/user-guide/data_model.rst]: Resolved in 01cec98b5f7c685e91bfaf611b3ecdf52833c072.

File: doc/user-guide/data_model.rst
Comment (akaszynski) [reply doc/user-guide/data_model.rst:51]: Added in e470ef90ba225493a7b260853cb0dab5f35446e8

File: doc/user-guide/data_model.rst
Comment (akaszynski) [reply doc/user-guide/data_model.rst]: Added most of this in 8e39b1ab1e70df7a748368c1865c36616ad3fbb7 (including figures, build docs to see).

image

I didn't add the entirety of your suggestion, so I'm leaving this unresolved for your second pass.

File: doc/user-guide/data_model.rst
Comment (akaszynski) [reply doc/user-guide/data_model.rst:201]: Good points.  Added this.

File: doc/user-guide/data_model.rst
Comment (akaszynski) [reply doc/user-guide/data_model.rst:239]: Added

File: doc/user-guide/data_model.rst
Comment (akaszynski) [reply doc/user-guide/data_model.rst]: Modified to read:

```
At the most fundamental level, all PyVista geometry classes inherit
from the :ref:`ref_dataset` class. A dataset has geometry, topology,
and attributes describing that geometry in the form of point, cell, or
field arrays.
```

**Comment by MatthewFlamm** [comment]
On the topic of the initial point of the PR 😄 :

> And I guess we should ask ourselves what this reversal of arguments really implies, why it makes sense, and write _that_ reasoning into a comment if necessary.

Using the issues code:

```python
import pyvista
import numpy

class Foo(pyvista.PolyData):
    pass

foo = Foo()
foo.points = np.eye(3)
foo.delaunay_2d(inplace=True)
```

It seems a little weird that we overwrite data from `PolyData` into an instance from an inherited class, in this case `foo`.  This only works if there is no functionality in the inherited class that is inconsistent with `PolyData`.  Looking at the overwrite function, it calls these two main methods:

https://github.com/pyvista/pyvista/blob/1f7b911df803c7d088d46b6de3a65ebfd3726edc/pyvista/core/dataset.py#L830-L844

https://github.com/pyvista/pyvista/blob/1f7b911df803c7d088d46b6de3a65ebfd3726edc/pyvista/core/dataobject.py#L44-L46

https://github.com/pyvista/pyvista/blob/1f7b911df803c7d088d46b6de3a65ebfd3726edc/pyvista/core/dataset.py#L589-L594

There are two kinds of logic here that I see.  Swapping out information in the underlying vtk objects, and then also modifying the texture information.  This still seems OK to me from compatibility, but I'm not a user of textures.

I agree that a comment would be useful here by distilling this down somehow.

**Comment by akaszynski** [comment]
> It seems a little weird that we overwrite data from PolyData into an instance from an inherited class, in this case foo. This only works if there is no functionality in the inherited class that is inconsistent with PolyData.

I think this is one of those cases where "Although practicality beats purity."  We want users to be able to subclass our classes and have it work "out of the box", which means that `inplace=True` should work as well.

In terms of compatibility, if they're subclassing something, they should expect some things to break and that's on them.  Since there are remarkably few methods/features that we use with `overwrite`, it doesn't seem like we need to add attribute checking to make sure this works.  That's their responsibility.

**Comment by MatthewFlamm** [comment]
I have (what I think is a better) proposal in that we allow developers to override the mapping in `pyvista.wrap`.  Then we can leave the `isinstance` check here unchanged.  Additional comments in #1565 has pointed out that the changes in this PR only address part of the issue when subclassing a dataset type.  See #1584.

**Comment by akaszynski** [comment]
> I have (what I think is a better) proposal in that we allow developers to override the mapping in pyvista.wrap. Then we can leave the isinstance check here unchanged. Additional comments in #1565 has pointed out that the changes in this PR only address part of the issue when subclassing a dataset type. See #1584.

I'll remove my changes here.  Even if #1584 falls through, we should have a separate PR for it (as this PR has morphed into a docs PR).

File: pyvista/core/datasetattributes.py
Comment (MatthewFlamm) [inline pyvista/core/datasetattributes.py:339]: I think this functionality of setting either scalar or vector, should be separated from whether it is set as the active scalar or vector.

There could be a keyword arg to choose, but it gets messy with the choices of active_scalars and active_vectors.  Can a vector data be set as active_scalars?

**Review by MatthewFlamm — COMMENTED** [review]
I haven't reviewed in detail yet.  The addition of functionality of vectors and scalars is fairly significant IMO.  We should consider whether the other parts of the code base need to be checked now.  For example, my understanding is that mesh rotations should rotate the vector data but not multi component scalar data.

https://vtk.org/Wiki/VTK/Tutorials/ScalarsVsVectors

**Comment by akaszynski** [comment]
> The addition of functionality of vectors and scalars is fairly significant IMO

I'm creating a new PR.  Our implementation of ``DataSetAttributes`` needs work and I've discovered this while going through the documentation.  @adeak and I have spoken about this before and it really needs to be overhauled in its own PR.  Splitting this off and I'll come back to this PR once that's merged.

**Comment by adeak** [comment]
I think Matthew was even more involved:
1. https://github.com/pyvista/pyvista/issues/1319
2. https://github.com/pyvista/pyvista/pull/1321

File: doc/user-guide/what-is-a-mesh.rst
Comment (adeak) [inline doc/user-guide/what-is-a-mesh.rst:139]: Merge malfunction

File: examples/01-filter/glyphs.py
Comment (adeak) [inline examples/01-filter/glyphs.py:24]: Another.

**Comment by akaszynski** [comment]
I think we might have one example plot too many for this to run on our CI/CD.

File: tests/test_datasetattributes.py
Comment (akaszynski) [inline tests/test_datasetattributes.py:20]:
```suggestion
```

File: tests/test_datasetattributes.py
Comment (akaszynski) [inline tests/test_datasetattributes.py:131]:
```suggestion
```

File: pyvista/core/datasetattributes.py
Comment (akaszynski) [reply pyvista/core/datasetattributes.py]: Resolved in #1593.

**Comment by akaszynski** [comment]
I've put other work on hold working on this PR and I'd really like to finish up #1557 and then get a release ready next week.  If either @adamgranthendry or @adeak want to submit a cleanup PR with suggestions on how to improve the documentation more, please feel free to submit one.

**Comment by adamgranthendry** [comment]
@akaszynski I apologize, I've been MIA these past few weeks on urgent work business. In full disclosure, I might not be successful in submitting a clean up PR this week until this business is completed.

As an aside, part of my work has required building VTK from source on Windows via CMake and Visual Studio and I have been successful at that. If not in this PR, I would like to add my completed instructions to the docs as I think it would be helpful to the community of Windows users out there.

Thank you @akaszynski and the rest of the PyVista team for all your hard work in this PR! It's great to see so much work being done!

**Comment by akaszynski** [comment]
> I would like to add my completed instructions to the docs as I think it would be helpful to the community of Windows users out there.

I've not yet been able to build VTK on my own, and this would be really helpful for PyVista (or VTK) users in general.

No rush on the cleanup; we release minor releases ever 6-8 weeks, but can submit patch releases as needed.

**Comment by adamgranthendry** [comment]
> I've not yet been able to build VTK on my own, and this would be really helpful for PyVista (or VTK) users in general.

As a sneak peek, see my SO post [Build VTK with Python Bindings and CUDA, TBB, & MPI on Windows](https://stackoverflow.com/questions/68929181/build-vtk-with-python-bindings-and-cuda-tbb-mpi-on-windows).

**Comment by adeak** [comment]
>  If either @adamgranthendry or @adeak want to submit a cleanup PR with suggestions on how to improve the documentation more, please feel free to submit one.

As I said in [this comment](https://github.com/pyvista/pyvista/pull/1571/#pullrequestreview-724934995), I meant to give this a quick proof-read before merging. I'll try to do the same in a follow-up PR within the next few days...

**Comment by akaszynski** [comment]
Sorry @adeak!

**Pull request merged**
